recipes/openai_terms.py

```diff
@@ -0,0 +1,379 @@
+
+    The results can then be corrected using the `prodigy textcat.manual` recipe and
+    turned into patterns via `prodigy terms.to-patterns`.
+    """
+    tic = time.time()
+    template = _load_template(prompt_path)
+    # The `best_of` param cannot be less than the amount we batch.
+    if best_of < n_batch:
+        best_of = n_batch
+
+    # Start collection of terms. If we resume we also fill seed terms with file contents.
+    terms = []
+    if resume:
+        if output_path.exists():
+            examples = srsly.read_jsonl(output_path)
+            terms.extend([e["text"] for e in examples])
+
+    # Mimic behavior from Prodigy terms recipe to ensure that seed terms also appear in output
+    for seed in seeds:
+        if seed not in terms:
+            srsly.write_jsonl(
+                output_path,
+                [{"text": seed, "meta": {"openai_query": query}}],
+                append=True,
+                append_new_line=False,
+            )
+
+    # Ensure we have access to correct environment variables and construct headers
+    if not os.getenv("OPENAI_KEY"):
+        msg.fail("The `OPENAI_KEY` is missing from your `.env` file.", exits=1)
+
+    if not os.getenv("OPENAI_ORG"):
+        msg.fail("The `OPENAI_ORG` is missing from your `.env` file.", exits=1)
+
+    headers = _generate_headers()
+
+    # This recipe may overshoot the target, but we keep going until we have at least `n`
+    while len(terms) < n:
+        prompt = template.render(n=n, examples=seeds + terms, description=query)
+        if verbose:
+            rich.print(Panel(prompt, title="Prompt to OpenAI"))
+
+        make_request = lambda: httpx.post(
+            "https://api.openai.com/v1/completions",
+            headers=headers,
+            json={
+                "model": model,
+                "prompt": [prompt],
+                "temperature": temperature,
+                "max_tokens": max_tokens,
+                "top_p": top_p,
+                "n": min(n_batch, best_of),
+                "best_of": best_of,
+            },
+            timeout=45,
+        )
+
+        # Catch 429: too many request errors
+        resp = _retry429(make_request, n=1, timeout_s=30)
+
+        # Report on any other error that might happen, the most typical use-case is
+        # catching the maximum context length of 4097 tokens when the prompt gets big.
+        if resp.status_code != 200:
+            msg.fail(f"Received status code {resp.status_code} from OpenAI. Details:")
+            rich.print(resp.json())
+            exit(code=1)
+
+        # Cast to a set to make sure we remove duplicates
+        choices = resp.json()["choices"]
+        sets_of_terms = [set(_parse_terms(c["text"])) for c in choices]
+        parsed_terms = list(reduce(lambda a, b: a.union(b), sets_of_terms))
+
+        # Save intermediate results into file, in-case of a hiccup
+        srsly.write_jsonl(
```

Comment:
I would either write to  the final file outside the loop trimming the output to `n` or document that the recipe will likely overshoot the `n` target.